exec/jdbc/src/main/java/org/apache/drill/jdbc/impl/DrillStatementImpl.java

```diff
@@ -38,8 +44,12 @@
 // methods for compatibility.)
 class DrillStatementImpl extends AvaticaStatement implements DrillStatement,
                                                              DrillRemoteStatement {
+  //Not using the DrillbitContext's ExecutorService as this is threadPool is light-weight (threads wake up to cancel tasks) but needs a low response time
+  private static ExecutorService queryTimeoutTaskPool = Executors.newCachedThreadPool(new NamedThreadFactory("q-timeout-"));
```

Comment:
DrillCursor is handling all the logic of executing queries, and waiting for results. It has access to the connection and the statement, so you would know the timeout (if set). In the cursor, we are using a lock for the first message, and a blocking queue for the batches, but when waiting on those, there's no timeout set. Instead we could use query timeout (or the remaining time left since the beginning of the execution) and throws SqlTimeoutException when the locks throws TimeoutException themselves.

In that scenario, no thread pool involved (except the one for I/O but it was already existing)